src/cdk/scrolling/virtual-scroll-viewport.ts

```diff
@@ -0,0 +1,189 @@
+  host: {
+    'class': 'cdk-virtual-scroll-viewport',
+  },
+  encapsulation: ViewEncapsulation.None,
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  preserveWhitespaces: false,
+})
+export class CdkVirtualScrollViewport implements OnInit, DoCheck, OnDestroy {
+  private _disconnectSubject = new Subject<void>();
+
+  private _renderedRangeSubject = new Subject<Range>();
+
+  /** The direction the viewport scrolls. */
+  @Input() orientation: 'horizontal' | 'vertical' = 'vertical';
+
+  /** The element that wraps the rendered content. */
+  @ViewChild('contentWrapper') _contentWrapper: ElementRef;
+
+  /** The total size of all content, including content that is not currently rendered. */
+  get totalContentSize() { return this._totalContentSize; }
+  set totalContentSize(size: number) {
+    if (this._totalContentSize != size) {
+      this._ngZone.run(() => {
+        this._totalContentSize = size;
+        this._changeDetectorRef.markForCheck();
+      });
+    }
+  }
+  private _totalContentSize = 0;
+
+  /** The currently rendered range of indices. */
+  get renderedRange() { return this._renderedRange; }
+  set renderedRange(range: Range) {
+    if (!this._rangesEqual(this._renderedRange, range)) {
+      this._ngZone.run(() => {
+        this._renderedRangeSubject.next(this._renderedRange = range);
+        this._changeDetectorRef.markForCheck();
+      });
+    }
+  }
+  private _renderedRange: Range = {start: 0, end: 0};
+
+  /** The offset of the rendered portion of the data from the start. */
+  get renderedContentOffset() { return this._renderedContentOffset; }
+  set renderedContentOffset(offset) {
+    if (this._renderedContentOffset != offset) {
+      this._ngZone.run(() => {
+        this._renderedContentOffset = offset;
+        this._renderedContentTransform = this.orientation === 'horizontal' ?
+            `translateX(${offset}px)`: `translateY(${offset}px)`;
+        this._changeDetectorRef.markForCheck();
+      });
+    }
+  }
+  private _renderedContentOffset = 0;
+
+  /** The length of the data connected to this viewport. */
+  get dataLength() { return this._dataLength; }
+  private _dataLength = 0;
+
+  /** The size of the viewport. */
+  get viewportSize() { return this._viewportSize; }
+  private _viewportSize = 0;
+
+  /** A stream that emits whenever the rendered range changes. */
+  renderedRangeStream: Observable<Range> = this._renderedRangeSubject.asObservable();
+
+  _renderedContentTransform: string;
+
+  private _connected = false;
+
+  private _scrollHandledStatus: 'needed' | 'pending' | 'done' = 'done';
+
+  private _scrollStrategyInited = false;
+
+  constructor(public elementRef: ElementRef, private _changeDetectorRef: ChangeDetectorRef,
+              private _ngZone: NgZone,
+              @Inject(VIRTUAL_SCROLL_STRATEGY) private _scrollStrategy: VirtualScrollStrategy) {}
+
+  /** Connect a `CdkForOf` to this viewport. */
+  connect(forOf: CdkForOf<any>) {
```

Comment:
This isn't formalized anywhere, but I feel that `attach` and `detach` are more appropriate for components that interact than `connect`/`disconnect`